core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -147,8 +147,6 @@ void handleUncaughtThrowable(Throwable t) {
   // Only null after channel is terminated. Must be assigned from the channelExecutor.
   private NameResolver nameResolver;
 
-  private final ProxyDetector proxyDetector;
```

Comment:
Note that we should plumb the `proxyDetector` eventually through here to the Name Resolver. We shouldn't pass the `proxyDetector` is the constructor like we do now. We could pass it in as `Attributes params`.

(CC @zhangkun83, I wonder if we should revisit that `params` argument to the NameResolver and do something more similar to Params/Helper from LB. That can be later though.)

Reply:
Will application provide ProxyDetector? Do we want NameResolvers to decide whether to use proxy or not?

Reply:
We want proxies to make the final determination, but we should provide the detector. We shouldn't assume such information is global. There's #3957 for example.